File: problems/41-first-missing-positive/math.cpp
/**
 * 答案范围为`[1, N]`，使用：抽屉原理
 * 
 * 时间：`O(N)`, 0ms
 * 空间：`O(1)`
 */

class Solution {
public:
    int firstMissingPositive(vector<int>& nums) {
        n = nums.size();

        for (int i = 0; i < n; ++i) {
            int num;
            while (isValid(num = nums[i]) && nums[num - 1] != num) {
                swap(nums, num - 1, i);
            }
        }

        // 第一个不合法（缺失）的数字
        int i;
        for (i = 0; i < n; ++i) {
            if (nums[i] != i + 1) {
                break;
            }
        }
        return i + 1;
    }

    bool isValid(int num) {
        return num >= 1 && num <= n;
    }

    void swap(vector<int> &nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }

private:
    int n;
};

File: 剑指offer/03/math.js
/**
 * 抽屉原理
 * 
 * 时间：O(N), 80ms
 * 空间：O(1), 40.5MB
 */
var findRepeatNumber = function (nums) {
  for (let i = 0; i < nums.length; ++i) {
    while (nums[nums[i]] !== nums[i]) { // 如果nums[i]不在正确位置上
      swap(nums, i, nums[i]);
    }
  }

  for (let i = 0; i < nums.length; ++i) {
    // 位置`i`上的数字不对，则返回该数字
    if (nums[i] !== i) {
      return nums[i];
    }
  }
};

// 在数组中交换两个位置
function swap(arr, i, j) {
  [arr[j], arr[i]] = [arr[i], arr[j]];
}